src/LoggerAnalyzer/LoggerAnalyzerAnalyzer.cs

```diff
@@ -33,42 +31,43 @@ public override void Initialize(AnalysisContext context)
-            if (genericParameterType is null)
-            {
-                continue;
-            }
-
-            if (genericParameterType.TypeArgumentList.Arguments.Count != 1)
-            {
-                continue;
-            }
-
-            var parameterSymbol = context.SemanticModel.GetDeclaredSymbol(parameter, context.CancellationToken);
-            if (parameterSymbol.Type.Name != "ILogger")
-            {
-                continue;
-            }
-
-            var classSymbol = context.SemanticModel.GetDeclaredSymbol(constructorDeclarationSyntaxNode.Parent, context.CancellationToken);
-
-            var genericArgument = genericParameterType.TypeArgumentList.Arguments.First() as IdentifierNameSyntax;
-            var genericArgumentType = context.SemanticModel.GetTypeInfo(genericArgument, context.CancellationToken);
-            if (genericArgumentType.Type.Name.Equals(classSymbol.Name, StringComparison.OrdinalIgnoreCase))
-            {
-                continue;
-            }
-
-            context.ReportDiagnostic(Diagnostic.Create(Rule, genericArgument.GetLocation(), classSymbol.Name));
-        }
+
+        context.RegisterCompilationStartAction(context =>
+        {
+            var loggerSymbol = context.Compilation.GetTypeByMetadataName("Microsoft.Extensions.Logging.ILogger`1");
+            if (loggerSymbol == null)
+                return;
+
+            context.RegisterSymbolAction(context =>
+            {
+                var symbol = (IMethodSymbol)context.Symbol;
+                if (symbol.MethodKind != MethodKind.Constructor)
+                    return;
+
+                foreach (var parameter in symbol.Parameters)
+                {
+                    if (parameter.Type is not INamedTypeSymbol parameterType)
+                        continue;
+
+                    // Only consider ILogger<T> parameters
+                    if (!parameterType.OriginalDefinition.Equals(loggerSymbol, SymbolEqualityComparer.Default))
+                        continue;
+
+                    // Check <T> is the containing class
+                    if (parameterType.TypeArguments[0].Equals(symbol.ContainingType, SymbolEqualityComparer.Default))
+                        continue;
+
+                    // Try to report the type argument. If not possible, report the parameter
+                    var syntax = parameter.DeclaringSyntaxReferences[0].GetSyntax();
+                    if (syntax is ParameterSyntax { Type: GenericNameSyntax { TypeArgumentList.Arguments: var arguments } } && arguments.Count == 1)
+                    {
+                        context.ReportDiagnostic(Diagnostic.Create(Rule, arguments[0].GetLocation(), symbol.ContainingType.Name));
+                        continue;
+                    }
+
+                    context.ReportDiagnostic(Diagnostic.Create(Rule, syntax.GetLocation(), symbol.ContainingType.Name));
+                }
+            }, SymbolKind.Method);

```

Comment:
Do I understand right it will check all methods but in the previous implementation, it checked only constructor declarations `context.RegisterSyntaxNodeAction(AnalyzeSyntax, SyntaxKind.ConstructorDeclaration);`. 
Is the new implementation more optimized?

Reply:
`if (symbol.MethodKind != MethodKind.Constructor)` quickly discards undesired methods.

> Is the new implementation more optimized?

I don't know if this is faster as it's up to Roslyn implementation details. Roslyn needs to compute symbols and will call the analyzer only when everything's ready. So it should be fast enough and there are lots of analyzer using this hook.